chainer/testing/math_function_test.py

```diff
@@ -0,0 +1,207 @@
+       returns another with the same ``dtype``.
+
+       >>> import unittest
+       >>> from chainer import testing
+       >>> from chainer import functions as F
+       >>>
+       >>> @testing.math_function_test(F.sin)
+       >>> class TestSin(unittest.TestCase):
+       >>>     pass
+
+       Because test methods are implicitly injected to ``TestSin`` class by the
+       decorator, we just place ``pass`` in the class definition.
+
+       We may use this decorator to test mathematical Chainer functions
+       implemented with composing other Chainer functions, like ``rsqrt``
+       which computes reciprocal of square root.
+
+       >>> import numpy
+       >>> import unittest
+       >>> from chainer import testing
+       >>> from chainer import functions as F
+       >>>
+       >>> def rsqrt(x, dtype=numpy.float32):
+       >>>     return numpy.reciprocal(numpy.sqrt(x, dtype=dtype))
+       >>>
+       >>> @testing.math_function_test(F.rsqrt, func_expected=rsqrt)
+       >>> class TestRsqrt(unittest.TestCase):
+       >>>     pass
+
+       Here we define ``rsqrt`` function composing numpy functions to get
+       expected values, passing it to ``func_expected`` keyword parameter of
+       ``@testing.math_function_test`` decorator.
+
+       We may also customize test data to be used. The following is an example
+       of testing ``sqrt`` Chainer function which we want to test in positive
+       value domain leaving some margin around zero of input ``x``.
+
+       >>> import numpy
+       >>> import unittest
+       >>> from chainer import testing
+       >>> from chainer import functions as F
+       >>>
+       >>> def make_data(dtype, shape):
+       >>>     x = numpy.random.uniform(0.1, 1, shape).astype(dtype)
+       >>>     gy = numpy.random.uniform(-1, 1, shape).astype(dtype)
+       >>>     return x, gy
+       >>>
+       >>> @testing.math_function_test(F.sqrt, make_data=make_data)
+       >>> class TestSqrt(unittest.TestCase):
+       >>>     pass
+
+       We define ``make_data`` function to return input and gradient ndarrays
+       generated in proper value domains with given ``dtype`` and ``shape``
+       parameters, then passing it to the decorator's ``make_data`` keyword
+       parameter.
+
+    """
+
+    # Import here to avoid mutual import.
+    from chainer import gradient_check
+    from chainer import testing
+
+    if func_expected is None:
+        name = func.__name__
+        try:
+            func_expected = getattr(numpy, name)
+        except AttributeError:
+            raise ValueError("numpy has no function corresponding "
+                             "to Chainer function '{}'.".format(name))
+
+    if label_expected is None:
+        label_expected = func.__name__
+    elif func_class(func) is None:
+        raise ValueError('Expected label is given even though Chainer '
+                         'function does not have its label.')
```

Comment:
As above, I intended to make it possible to test functions that have no `chainer.Function` subclass. Here I mean:

- If label_expected is not provided, label test may be skipped if `func` has no corresponding `chainer.Function` subclass.
- If label_expected is provided, it means that we want to test label so we need `func` to have corresponding `chainer.Function` subclass, so this error checking works.